### PR TITLE
[JENKINS-58750] Feature/build comment

### PR DESCRIPTION
**Pull request by baymac**
* Added new trait to log build status as comment for a commit or merge request
Screenshot from 2019-08-12 17-11-32
Screenshot from 2019-08-12 17-11-37

### CODE REVIEW

**Comment by casz** [comment]
This should support using a sudo token. With the sudo token you can make it appear as a different user.

Let's say I set up a token for our local Administrator user but want to post as our Bot user, I would you use Sudo.
https://docs.gitlab.com/ee/api/#sudo

**Comment by casz** [comment]
the config should not demand a secondary token but expect the server to have a single token with api and sudo permissions.
The sudo api returns a reasonable error if sudo token is not provided.

The config I think should be an optionalProperty (data binding ❤️)

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import hudson.Extension;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class LogCommentTrait extends SCMSourceTrait {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java]: LogCommentTrait should have a jelly config and databoundsetter for sudo :)

Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java]: Fixed in 883df7f

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java [hunk 2]
```diff
@@ -1,31 +1,43 @@
...
+    @NonNull
+    private String sudoUser;
+
     /**
      * Constructor for stapler.
      */
     @DataBoundConstructor
-    public LogCommentTrait() {
-        //empty
+    public LogCommentTrait(@Nonnull String sudoUser) {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java]: Please use `@DataBoundSetter` instead
I don't think it should be mandatory.

Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/LogCommentTrait.java]: Fixed in c56e59a

**Comment by baymac** [comment]
@casz did you test comment logging as sudo user? I am unable to due to my server issues but when I tested the api separately it seemed to be working. Let me know once you are able to verify that it is working, we can merge this.

**Comment by casz** [comment]
Perhaps there should be more option? Only log failure? :sweat_smile:

**Comment by baymac** [comment]
Can be a little specific about the "more option"? 😅

**Comment by casz** [comment]
Like only comment on failure?

**Comment by baymac** [comment]
@casz Okay I will add a few strategies.

**Comment by baymac** [comment]
I decided to go for a checkbox if user wants to log successful builds or not.

**Comment by baymac** [comment]
The new trait makes it compatible with Folder organization as well! But the `isMatch` triggers the build twice for some reason. :sweat:

**Pull request merged**